File: app/src/main/java/com/salino/sali/ui/screens/activityfeed/ActivityFeedViewModel.kt
package com.salino.sali.ui.screens.activityfeed

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.salino.sali.data.model.ActivityLog
import com.salino.sali.domain.repository.ActivityRepository
import com.salino.sali.domain.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ActivityFeedState(
    val entries: List<ActivityLog> = emptyList(),
    val isLoading: Boolean = true
)

@HiltViewModel
class ActivityFeedViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val activityRepository: ActivityRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(ActivityFeedState())
    val uiState: StateFlow<ActivityFeedState> = _uiState

    init {
        viewModelScope.launch {
            authRepository.observeCurrentUser().filterNotNull().collect { user ->
                val householdId = user.activeHouseholdId ?: return@collect
                activityRepository.observeActivityFeed(householdId).collect { entries ->
                    _uiState.value = ActivityFeedState(entries = entries, isLoading = false)
                }
            }
        }
    }
}

File: ios/ViewModels/ActivityFeedViewModel.swift
import Foundation
import Combine

struct ActivityFeedState {
    var entries: [ActivityLog] = []
    var isLoading = true
}

@MainActor
final class ActivityFeedViewModel: ObservableObject {
    @Published var state = ActivityFeedState()
    private let authRepository: AuthRepository
    private let activityRepository: ActivityRepository
    private var task: Task<Void, Never>?

    init(authRepository: AuthRepository, activityRepository: ActivityRepository) {
        self.authRepository = authRepository
        self.activityRepository = activityRepository
        task = Task { await observeActivity() }
    }

    deinit { task?.cancel() }

    private func observeActivity() async {
        guard let user = await firstCurrentUser(from: authRepository), let householdId = user.activeHouseholdId else {
            state.isLoading = false
            return
        }
        for await entries in activityRepository.observeActivityFeed(householdId: householdId) {
            state = ActivityFeedState(entries: entries, isLoading: false)
        }
    }
}
